src/libraries/System.Linq.Expressions/src/System/Runtime/CompilerServices/ReadOnlyCollectionBuilder.cs

```diff
@@ -451,7 +451,7 @@ private static bool IsCompatibleObject(object? value)
 
         private static void ValidateNullValue(object? value, string argument)
         {
-            if (value == null && default(T) != null)
+            if (value == null && typeof(T).IsValueType)
```

Comment:
This will be wrong for `Nullable<T>`, e.g. if value is null and typeof(T) is typeof(int?), then the code previously wouldn't have thrown but now it will.